lib/modules/manager/nix/extract.ts

```diff
@@ -1,23 +1,139 @@
+import { logger } from '../../../logger';
+import { getSiblingFileName, readLocalFile } from '../../../util/fs';
 import { regEx } from '../../../util/regex';
 import { GitRefsDatasource } from '../../datasource/git-refs';
-import { id as nixpkgsVersioning } from '../../versioning/nixpkgs';
 import type { PackageDependency, PackageFileContent } from '../types';
+import { NixFlakeLock } from './schema';
 
-const nixpkgsRegex = regEx(/"github:nixos\/nixpkgs(\/(?<ref>[a-z0-9-.]+))?"/i);
+// TODO: add support to update nixpkgs branches in flakes.nix using nixpkgsVersioning
+
+// as documented upstream
+// https://github.com/NixOS/nix/blob/master/doc/manual/source/protocols/tarball-fetcher.md#gitea-and-forgejo-support
+const lockableHTTPTarballProtocol = regEx(
+  '^https://(?<domain>[^/]+)/(?<owner>[^/]+)/(?<repo>[^/]+)/archive/(?<rev>.+).tar.gz$',
+);
+
+export async function extractPackageFile(
+  content: string,
+  packageFile: string,
+): Promise<PackageFileContent | null> {
+  const packageLockFile = getSiblingFileName(packageFile, 'flake.lock');
+  const lockContents = await readLocalFile(packageLockFile, 'utf8');
+
+  logger.trace(`nix.extractPackageFile(${packageLockFile})`);
 
-export function extractPackageFile(content: string): PackageFileContent | null {
   const deps: PackageDependency[] = [];
 
-  const match = nixpkgsRegex.exec(content);
-  if (match?.groups) {
-    const { ref } = match.groups;
-    deps.push({
-      depName: 'nixpkgs',
-      currentValue: ref,
-      datasource: GitRefsDatasource.id,
-      packageName: 'https://github.com/NixOS/nixpkgs',
-      versioning: nixpkgsVersioning,
-    });
+  const flakeLockParsed = NixFlakeLock.safeParse(lockContents);
+  if (!flakeLockParsed.success) {
+    logger.debug(
+      { packageLockFile, error: flakeLockParsed.error },
+      `invalid flake.lock file`,
+    );
+    return null;
+  }
+
+  const flakeLock = flakeLockParsed.data;
+
+  for (const depName of Object.keys(flakeLock.nodes)) {
```

Comment:
```suggestion
  for (const [depName, flakeInput] of Object.entries(flakeLock.nodes)) {
```
you can have key and value at once, so you don't need to fetch later.

Reply:
not a js/ts developer, thanks